backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ir/interop/cenum/CEnumClassGenerator.kt

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+package org.jetbrains.kotlin.backend.konan.ir.interop.cenum
+
+import org.jetbrains.kotlin.backend.konan.InteropBuiltIns
+import org.jetbrains.kotlin.backend.konan.descriptors.enumEntries
+import org.jetbrains.kotlin.backend.konan.ir.interop.DescriptorToIrTranslationMixin
+import org.jetbrains.kotlin.backend.konan.ir.interop.findDeclarationByName
+import org.jetbrains.kotlin.descriptors.ClassDescriptor
+import org.jetbrains.kotlin.descriptors.PropertyDescriptor
+import org.jetbrains.kotlin.descriptors.Visibilities
+import org.jetbrains.kotlin.ir.builders.*
+import org.jetbrains.kotlin.ir.declarations.*
+import org.jetbrains.kotlin.ir.descriptors.IrBuiltIns
+import org.jetbrains.kotlin.ir.expressions.IrExpression
+import org.jetbrains.kotlin.ir.expressions.impl.IrEnumConstructorCallImpl
+import org.jetbrains.kotlin.ir.expressions.impl.IrInstanceInitializerCallImpl
+import org.jetbrains.kotlin.ir.util.*
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.name.Name
+import org.jetbrains.kotlin.psi2ir.generators.DeclarationGenerator
+import org.jetbrains.kotlin.psi2ir.generators.EnumClassMembersGenerator
+import org.jetbrains.kotlin.psi2ir.generators.GeneratorContext
+import org.jetbrains.kotlin.resolve.constants.ConstantValue
+import org.jetbrains.kotlin.utils.addToStdlib.firstNotNullResult
+
+internal class CEnumClassGenerator(
+        val context: GeneratorContext,
+        interopBuiltIns: InteropBuiltIns
+) : DescriptorToIrTranslationMixin {
+
+    override val irBuiltIns: IrBuiltIns = context.irBuiltIns
+    override val symbolTable: SymbolTable = context.symbolTable
+    override val typeTranslator: TypeTranslator = context.typeTranslator
+
+    companion object {
+        private val cEnumEntryValueAnnotationName = FqName("kotlinx.cinterop.internal.CEnumEntryValue")
+        private val cEnumEntryValueTypes = setOf(
+                "Byte", "Short", "Int", "Long",
+                "UByte", "UShort", "UInt", "ULong"
+        )
+    }
+
+    private val enumClassMembersGenerator = EnumClassMembersGenerator(DeclarationGenerator(context))
+
+    /**
+     * Searches for an IR class for [classDescriptor] in symbol table.
+     * Generates one if absent.
+     */
+    fun findOrGenerateCEnum(classDescriptor: ClassDescriptor, parent: IrDeclarationContainer): IrClass {
+        val irClassSymbol = symbolTable.referenceClass(classDescriptor)
+        return if (!irClassSymbol.isBound) {
+            provideIrClassForCEnum(classDescriptor).also {
+                it.patchDeclarationParents(parent)
```

Comment:
Consider passing parent to the building function. Usage of `patchDeclarationParents` is the last resort IMO.